Contribution guide (#7)
tonic-team/tonic.site: CONTRIBUTING.md

# Contributing to the Tonic site

Thank you for taking the time to contribute to the Tonic site.

> Contributions are welcome and this document is intended to help them to find their way into the project.

Reviewer: ```suggestion
Contributions are welcome and this document is intended to help you find your way into the project.
Please do not be afraid by its length. 
While we will appreciate if your pull request contains correctly styled code,
we will be willing to clean your pull requests ourselves.
This guide indeed contains a lot of information about code styling for advanced users and serves as a documentation tool for the core contribution team.
```

Author: Hm, I get your point, but it sounds quite defensive to me. I'll have to think this over a bit.

Author: I've added most of your suggestion because I thought of a cultural difference between scientific and open source communities. In open source projects, it is much more natural that everyone is invited to participate, whereas in science, this is no matter of course and a more explicit invitation might be appropriate. I have left out the more apologizing parts, because I think apologizing for length doesn't make it shorter. ;-)  And ironically, if I had targeted only advanced users, the guide would have become remarkably shorter.

## What can be contributed?

There are several ways to contribute:

- Ideas, suggestions, issue reports
- smaller fixes on existing pages
- Content: new pages or larger additions to existing pages

If you are missing something or have discovered incorrect informations,
you can use the project's [issue tracker] to briefly describe what you'd like to see on the site.
Fixes or content can be submitted via fork and pull request on Github.

## Minor improvements via browser

> It is possible and absolutely fine to provide smaller fixes like typo corrections and similar suggestions in a pure browser-based workflow.

Reviewer: ```suggestion
It is possible and absolutely fine to provide fixes and suggestions in a pure browser-based workflow.
```

If you hit the _edit this page_ button on a website page, you will be redirected to the respective edit page on Github,
where you can improve the Markdown source, create a commit and submit it as a pull request.
If you don't have write access to the repo and if you are a first-time contributor, you will be asked to create a fork before you can edit the page file.

> 

Reviewer: ### code styling in the browser

## Contributions via local setup

The browser-based workflow might be sufficient for smaller improvements, but for larger contributions,
especially if you want to add new pages or intend to contribute for multiple times,
the recommended way is cloning the forked project to your machine and preparing your changes locally in a suited text editor.
This makes it easier to write clean files and to preview the result in the local development server before any PR is submitted.

### Workflow overview

This actually applies to the browser-based workflow, except that some steps in the browser are shortened a bit.
And of course both can be combined, e.g., one can open a PR in the browser but finish it locally.

1. Setup
   - Make sure you have [git] installed
   - If you want to preview the site in a local development server, install [hugo].
   - If you want to edit R Markdown files, you'll also need [R] and the [blogdown] package.
   - If you don't have write access to the project repo, fork it.
   - Clone the repo you want to write to
   - Create a new local branch from the main branch, name it after your contribution
2. Uploading your Changes (iterative)
   - Make your improvements

>    - Please try to write one sentence or clause per line, as far as it makes sense. This improves git diff and file reviews.

Reviewer: ```suggestion
   - Please try to write one sentence or clause per line, as far as it makes sense. This improves `git diff`(previews of differences) and file reviews.
```

   - Please leave the “allow edits” checkbox enabled
4. Checking, cleaning (iterative)
   - When a PR is created, Prettier runs on the PR branch.
     If it detects unformatted files, a commit that applies formatting changes is added to the branch.
     if you want to add more commits to the PR, you will have to pull these changes into your local branch.
   - After prettifying is done, the Markdown files are checked for issues that relate to ambiguity, accessibility, consistency, and alike by [markdownlint].
     If this check doesn't pass, you have to add commits to your branch which fix the detected issues.
   - The checks rerun and the cycle repeats, until the checks succeed.
5. Reviewing and improving content (iterative)
   - Now your PR is ready for review, and you can request review from one or more collaborators.
   - If a reviewer rejects your PR and provides feedback, you will have to discuss or integrate the feedback
     by adding fixup commits to your PR branch until it gets approved.
     This also triggers the checking and cleaning cycle.
   - Meanwhile, if the main branch receives new commits and your PR can't be merged anymore due to merge conflicts,

>      please don't merge main into your branch, but rebase your branch on main and force-push your branch.

Reviewer: this needs explanation or a link.

Author: Added this link: <https://www.atlassian.com/git/tutorials/merging-vs-rebasing>

     This keeps the commit history linear.
     - Finally, a reviewer might approve your PR
6. Merge
   - The PR is clean, approved, and thus ready for merge.
   - A maintainer will merge it.

### Content file structure

The content on this site is authored in Markdown, a lightweight markup language that is rendered to HTML.
Page files are written in markdown or R markdown and organized in the content folder.
The folder names represent the URL segments, and the menu is built from the folder structure.

- Chapter pages are located at `content/<chapter-name>/_index.md`.
  The underscore is important, and the metadata block should contain `chapter: true`.
  Note that the text will be centered when `chapter:` is set to `true`.
- Child pages are located either at `content/<chapter-name>/<page-name>/index.md` or at `content/<chapter-name>/<page-name>.md`.

>   If the former variant is used (page bundles), static assets can put directly into the page directory.

Reviewer: please define or examplify static assets.

Author: Thanks, I've added some explanations and examples illustrating the usage of static assets.

>   If the former variant is used (page bundles), static assets can put directly into the page directory.

Reviewer: maybe this should be moved at the start of the document (for beginners to read).

Author: Indeed, done.

# Contributing to the Tonic site

Thank you for taking the time to contribute to the Tonic site.
Contributions are welcome and this document is intended to help you find your way into the project.
While we will appreciate if your pull request contains correctly styled code,
we will be willing to clean your pull requests ourselves.

## What can be contributed?

There are several ways to contribute:

- Ideas, suggestions, issue reports

> - smaller fixes on existing pages

Reviewer 2: ```suggestion
- Smaller fixes on existing pages
```

Capitalise for consistency?

- Content: new pages or larger additions to existing pages

> If you are missing something or have discovered incorrect informations,

Reviewer 2: ```suggestion
If you are missing something or have discovered incorrect information,
```

Consult the [theme documentation] for more information about content editing.

## Minor improvements via browser

It is possible and absolutely fine to provide fixes and suggestions in a pure browser-based workflow.
If you hit the _edit this page_ button on a website page, you will be redirected to the respective edit page on Github,
where you can improve the Markdown source, create a commit and submit it as a pull request.
If you don't have write access to the repo and if you are a first-time contributor, you will be asked to create a fork before you can edit the page file.

In this project, we strive to maintain a clean and consistent code style.
the files in the PR are formatted by [prettier], a code formatting tool which understands many languages,
but you can use the [prettier playground] to preformat your markdown Before you commit and submit the PR:

1. Make your changes
2. Paste the full document into the input field of the playground (left side)
3. Replace document with Prettier's result from the output field (right side).
4. Commit and submit

Some quality checks will run on the PR before it is ready for review.

> This takes a moment, and you will probably be informed per email if your PR introduces an issue.

Reviewer 2: ```suggestion
This takes a moment, and you will probably be informed via email if your PR introduces an issue.
```

You can fix the reported issues by adding more commits to the PR.
It's relatively unlikely to introduce an error in a small change, but not impossible.

> The Markdown files are checked using [markdownlint], a tool which helps to discaver errors and inconsistencies in Markdown files..

Reviewer 2: ```suggestion
The Markdown files are checked using [markdownlint], a tool which helps to discover errors and inconsistencies in Markdown files..
```

You can use its browser-based [online demo][markdownlint-demo] to inspect the file you have edited and figure out what went wrong.

## Contributions via local setup

The browser-based workflow might be sufficient for smaller improvements, but for larger contributions,

> especially if you want to add new pages or intend to contribute for multiple times,

Reviewer 2: ```suggestion
especially if you want to add new pages or intend to contribute multiple times,
```

> the recommended way is cloning the forked project to your machine and preparing your changes locally in a suited text editor.

Reviewer 2: ```suggestion
the recommended way is cloning the forked project to your machine and preparing your changes locally in a suitable text editor.
```

This makes it easier to write clean files and to preview the result in the local development server before any PR is submitted.

### Workflow overview

This actually applies to the browser-based workflow, except that some steps in the browser are shortened a bit.
And of course both can be combined, e.g., one can open a PR in the browser but finish it locally.

1. Setup
   - Make sure you have [git] installed
   - If you want to preview the site in a local development server, install [hugo].
   - If you want to edit R Markdown files, you'll also need [R] and the [blogdown] package.
   - If you don't have write access to the project repo, fork it.
   - Clone the repo you want to write to
   - Create a new local branch from the main branch, name it after your contribution
2. Uploading your Changes (iterative)
   - Make your improvements
   - Please try to write one sentence or clause per line, as far as it makes sense. This improves `git diff`(previews of differences) and file reviews.
   - Try to format your edited files with [Prettier] and lint Markdown with [markdownlint].

>    - commit to the new branch.

Reviewer 2: ```suggestion
   - Commit to the new branch.
```
